packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+  });
+
+  /// The magnifier's opacity.
+  final double opacity;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+      return true;
+    }
+
+    return super == other && other is MagnifierDecoration && other.opacity == opacity;
+  }
+
+  @override
+  int get hashCode => Object.hash(super.hashCode, opacity);
+}
+
+/// A common building base for magnifiers.
+///
+/// {@template flutter.widgets.magnifier.intro}
+/// This magnifying glass is useful for scenarios on mobile devices where
+/// the user's finger may be covering part of the screen where a granular
+/// action is being performed, such as navigating a small cursor with a drag
+/// gesture, on an image or text.
+/// {@endtemplate}
+///
+/// A magnifier can be convienently managed by [MagnifierController], which handles
+/// showing and hiding the magnifier, with an optional entry / exit animation.
+///
+/// See:
+/// * [MagnifierController], a controller to handle magnifiers in an overlay.
+class RawMagnifier extends StatelessWidget {
+  /// Constructs a [RawMagnifier].
+  ///
+  /// {@template flutter.widgets.magnifier.RawMagnifier.invisibility_warning}
+  /// By default, this magnifier uses the default [MagnifierDecoration],
+  /// the focal point is directly under the magnifier, and there is no magnification:
+  /// This means that a default magnifier will be entirely invisible to the naked eye,
+  /// since it is painting exactly what is under it, exactly where it was painted
+  /// orignally.
+  /// {@endtemplate}
+  const RawMagnifier({
+      super.key,
+      this.magnificationScale = 1,
+      required this.size,
+      this.focalPoint = Offset.zero,
+      this.child,
+      this.decoration = const MagnifierDecoration()
+      }) : assert(magnificationScale != 0,
+            'Magnification scale of 0 results in undefined behavior.');
+
+  /// This magnifier's decoration.
+  ///
+  /// {@macro flutter.widgets.magnifier.RawMagnifier.invisibility_warning}
+  final MagnifierDecoration decoration;
+
+  /// The size of the magnifier.
+  ///
+  /// This does not include added border; it only includes
+  /// the size of the magnifier.
+  final Size size;
+
+  /// The offset of the magnifier from [RawMagnifier]'s center.
+  ///
+  /// If [focalPoint] is [Offset.zero], then the [focalPoint]
+  /// will point directly below this [RawMagnifier].
+  final Offset focalPoint;
+
+  /// An optional widget to posiiton inside the len of the [RawMagnifier].
+  ///
+  /// This is positioned over the [RawMagnifier] - it may be useful for tinting the
+  /// [RawMagnifier], or drawing a crosshair like UI.
+  final Widget? child;
+
+  /// How "zoomed in" the magnification subject is in the lens.
+  final double magnificationScale;
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      clipBehavior: Clip.none,
+      alignment: Alignment.center,
+      children: <Widget>[
+          Opacity(
```

Comment:
nit: the indentation here seems to be off

Reply:
fixed w/ autoformatter